mesonbuild/compilers/detect.py

```diff
@@ -216,7 +216,13 @@ def detect_static_linker(env: 'Environment', compiler: Compiler) -> StaticLinker
         if "xilib: executing 'lib'" in err:
             return linkers.IntelVisualStudioLinker(linker, getattr(compiler, 'machine', None))
         if '/OUT:' in out.upper() or '/OUT:' in err.upper():
-            return linkers.VisualStudioLinker(linker, getattr(compiler, 'machine', None))
+            # It is using the VS-style linker/archiver syntax but it's important to
+            # distinguish between 'lib' and 'llvm-lib', which have incompatibilities
+            # (notably, when using LTO).
+            if 'LLVM' in out:
+                return linkers.LlvmVisualStudioLinker(linker, getattr(compiler, 'machine', None))
+            else:
+                return linkers.VisualStudioLinker(linker, getattr(compiler, 'machine', None))
```

Comment:
This change could also be a separate commit to reduce the diff since this look like a bug fix. Same with the similar chunk below.

Reply:
It's not a bug fix; it's a change that's needed in order to perform the 'mixing toolchain streams' validation done by `_validate_tool_combination` in interpreter.py.  Previously, use of both 'lib' or 'llvm-lib' would end up instantiating a VisualStudioLinker, after which, we can no longer distinguish the difference by normal means (i.e. 'id').  I think we'd need to open it up and inspect other members to figure out which flavour we're dealing with.